File: Dynamic_Programming/940.Distinct-Subsequences-II/940.Distinct-Subsequences-II.py
class Solution:
    def distinctSubseqII(self, S):
        """
        :type S: str
        :rtype: int
        """
        last = [0 for _ in range(26)]                
        dp = 0
        M = 1e9+7
        
        for ch in S:
            dp_temp = dp            
            dp = dp_temp + (dp_temp +1 - last[ord(ch)-ord('a')])
            dp %= M
            last[ord(ch)-ord('a')] = dp_temp+1
            last[ord(ch)-ord('a')] %= M
        return int(dp)

File: Dynamic_Programming/940.Distinct-Subsequences-II/940.Distinct-Subsequences-II.cpp
class Solution {
    long dp[2005];
    long M = 1e9+7;
public:
    int distinctSubseqII(string s) 
    {
        int n = s.size();
        s = "#"+s;
        vector<int>last(26,0);
        dp[0] = 1;

        for (int i=1; i<=n; i++)
        {
            int j = last[s[i]-'a'];
            dp[i] = dp[i-1] *2 %M - (j>=1 ? dp[j-1] : 0);
            dp[i] = (dp[i]+M) % M;
            last[s[i]-'a'] = i;
        }

        return (dp[n] -1 +M) % M;        
    }
};
